corehq/apps/userreports/README.md

```diff
@@ -385,6 +385,85 @@ Variables can be any valid numbers (Python datatypes `int`, `float`, and `long`
 
 More examples can be found on practical [examples page](corehq/apps/userreports/examples/examples.md#evaluator-examples).
 
+#### Case --> Form Lookup
+
+```json
+{
+   "type": "get_case_forms",
+   "case_id_expression": {},  # required
+   "filter": {},  # optional - possibly to remove for now in favor of filter expression
+}
+```
+
+Gets a list of forms from a case, optionally filtering (TBD if that belongs here).
+
+#### Filter Expression
+
+```json
+{
+  "type": "filter",  # not sure on name
+  "items_expression": {},  # expression should return a list
+  "filter_expression": {}  # will be called on each item, should return true or false indicating whether to include
+}
+```
+
+This should work just like python's `filter` function.
+
+#### Map Expression
+
+```json
+{
+  "type": "map",  # not sure on name
+  "items_expression": {},  # expression should return a list
+  "map_expression": {}  # expression to be evaluated on each item in the list
+}
+```
+
+In python the equivalent would be `map(lambda x: x.y, my_list)`, which would return a new list with the lambda function applied.
+
+#### Reduce Expression
+
+```json
+{
+  "type": "reduce",  # not sure on name
+  "items_expression":  {},  # expression should return a list
+  "aggregation_fn": "sum"  # one of a fixed set of allowed aggregation functions
+}
+```
+
+It would be cool if we could do custom reduces (by defining a function that takes in two values and spits out a single value). That might look like the following:
+
+```json
+{
+  "type": "custom_reduce",  # not sure on name
+  "items_expression":  {},  # expression should return a list
+  "reduce_statement": "x + y"  # required to only use these two variables
+}
+```
+
+#### Lookup Expression
```

Comment:
sorry I was assuming a first/last reduce function too. so I think you can do one of the two options below. The first one without map is actually cleaner. (let me know if I'm not understanding what you're trying to do correctly, and maybe you can write it in pseudo-python so it's easier to discuss)

```python
sorted(case_forms, key=lambda x: x.received_on).first().weight
map(lambda f: f.weight, sorted(case_forms, key=lambda x: x.received_on)).first()
```